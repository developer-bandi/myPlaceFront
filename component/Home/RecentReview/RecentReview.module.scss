.mainBlock {
  width: 100%;
  margin: 50px 0px 0px 0px;
}

.subBlock {
  width: 1200px;
  margin: 0px auto 0px auto;
  display: flex;
  justify-content: space-between;
  @media screen and (max-width: 1200px) {
    width: 100%;
  }
  @media screen and (max-width: 767px) {
    display: block;
  }
}

.reviewListContainer {
  width: calc(100% - 400px);
  height: 600px;
  overflow: auto;
  @media screen and (max-width: 1200px) {
    margin: 0px 20px 0px 0px;
  }
  @media screen and (max-width: 767px) {
    margin: 0px 20px 0px 20px;
    width: calc(100% - 40px);
  }
}

.reviewList {
  position: relative;
  height: 1925px;
  width: 100%;
  overflow: hidden;
}

.reviewListLoading {
  width: 100%;
  height: 60px;
  display: flex;
  justify-content: center;
  align-items: center;
  font-weight: 600;
  font-size: 25px;
}

.lastCheckbox {
  width: 100%;
  height: 192px;
  background-color: black;
  position: absolute;
  bottom: 0;
}

.error {
  text-align: center;
  font-size: 30px;
  margin: 50px 0 200px 0;
}
