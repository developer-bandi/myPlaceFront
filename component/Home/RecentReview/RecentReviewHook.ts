import {useEffect, useState} from "react";
import {axiosGetRecentReview} from "../../../lib/commonFn/api";

export interface recentReviewData {
  id: number;
  content: string;
  createdAt: string;
  storeName: string;
  storeAddress: string;
  storeLatitude: string;
  storeLongitude: string;
  nickname: string;
  hashtag: string[];
}

export interface reviewRecentState {
  content?: {count: number; rows: recentReviewData[]};
  loading: boolean;
  error: boolean;
}

const useRecentReview = () => {
  const [recentReviewData, setRecentReviewData] = useState<reviewRecentState>({
    loading: true,
    error: false,
  });
  useEffect(() => {
    const asyncTempFunc = async () => {
      try {
        const content = await axiosGetRecentReview();
        setRecentReviewData({
          content: content.data,
          loading: false,
          error: false,
        });
      } catch (err) {
        setRecentReviewData({loading: false, error: true});
      }
    };
    asyncTempFunc();
  }, []);

  return {recentReviewData};
};

export default useRecentReview;
