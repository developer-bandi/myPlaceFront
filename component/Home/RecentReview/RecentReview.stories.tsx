import { ComponentStory, ComponentMeta } from "@storybook/react";
import RecentReview from "./RecentReview";
import { action } from "@storybook/addon-actions";

export default {
  title: "Home/RecentReview",
  component: RecentReview,
  parameters: {
    layout: "fullscreen",
  },
} as ComponentMeta<typeof RecentReview>;

const Template: ComponentStory<typeof RecentReview> = (arg) => {
  return <RecentReview {...arg} />;
};

export const isLoading = Template.bind({});
isLoading.args = {
  serverData: { loading: true, error: false },
  moveTargetStore: action("moveTargetStore"),
};

export const isError = Template.bind({});
isError.args = {
  serverData: { loading: false, error: true },
  moveTargetStore: action("moveTargetStore"),
};

export const isNormal = Template.bind({});
isNormal.args = {
  serverData: {
    content: {
      count: 15,
      rows: new Array(10).fill(0).map((data, index) => {
        return {
          id: index + 1,
          content: "testContent",
          createdAt: "testCreatedAt",
          storeName: "testStoreName",
          storeAddress: "testStoreAddress",
          storeLatitude: "testStoreLatitude",
          storeLongitude: "testStoreLongitude",
          nickname: "testNickname",
          hashtag: ["testHashtag"],
        };
      }),
    },
    loading: false,
    error: false,
  },
  moveTargetStore: action("moveTargetStore"),
};
