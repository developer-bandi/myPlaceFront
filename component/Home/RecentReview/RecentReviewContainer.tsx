import {axiosGetRecentReview} from "../../../lib/commonFn/api";
import useGetServerData from "../../../lib/customHook/getData";
import useMoveTargetStore from "../../../lib/customHook/moveTargetStore";
import RecentReview from "./RecentReview";

export interface recentReviewData {
  id: number;
  content: string;
  createdAt: string;
  storeName: string;
  storeAddress: string;
  storeLatitude: string;
  storeLongitude: string;
  nickname: string;
  hashtag: string[];
}

export interface reviewRecentState {
  content?: {count: number; rows: recentReviewData[]};
  loading: boolean;
  error: boolean;
}

const RecentReviewContainer = () => {
  const {serverData} = useGetServerData(axiosGetRecentReview);
  const {moveTargetStore} = useMoveTargetStore();
  return (
    <RecentReview
      serverData={serverData as reviewRecentState}
      moveTargetStore={moveTargetStore}
    />
  );
};

export default RecentReviewContainer;
