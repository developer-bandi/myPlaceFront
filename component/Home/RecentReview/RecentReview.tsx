import {reviewRecentState} from "./RecentReviewHook";
import styles from "./RecentReview.module.scss";
import {setDateLatest} from "../../../lib/commonFn/date";

interface RecentReviewProps {
  recentReviewData: reviewRecentState;
  moveTargetStore: (
    id: number,
    name: string,
    latitude: string,
    longitude: string,
    address: string
  ) => void;
}

const RecentReview = ({
  recentReviewData,
  moveTargetStore,
}: RecentReviewProps) => {
  if (recentReviewData.loading) {
    return (
      <section className={styles.mainBlock} data-testid="loading">
        <div className={styles.subBlock}>
          <div className={styles.leftBlock}>
            <div className={styles.title}>
              <div className={styles.loadingTitle} />
              <div className={styles.loadingTitle} />
            </div>
            <div className={styles.summary}>
              <div className={styles.loadingSummary} />
              <div className={styles.loadingSummary} />
              <div className={styles.loadingSummary} />
              <div className={styles.loadingSummary} />
            </div>
          </div>
          <div className={styles.reviewList}>
            {new Array(3).fill(0).map((data, index) => {
              return (
                <div className={styles.reviewBlock} key={index}>
                  <div className={styles.loadingUser} />
                  <div className={styles.loadingStoreName} />
                  <div className={styles.loadingReview} />
                  <div className={styles.loadingReview} />
                </div>
              );
            })}
          </div>
        </div>
      </section>
    );
  } else if (recentReviewData.error) {
    return (
      <section className={styles.mainBlock} data-testid="error">
        <p className={styles.error}>에러가 발생하였습니다</p>
      </section>
    );
  } else if (recentReviewData.content !== undefined) {
    return (
      <section className={styles.mainBlock} data-testid="result">
        <div className={styles.subBlock}>
          <div className={styles.leftBlock}>
            <h3 className={styles.title}>
              현재
              <span className={styles.color}>
                {recentReviewData.content.count}
              </span>
              개의
              <br /> 리뷰가 등록되어 있습니다.
            </h3>
            <p className={styles.summary}>
              리뷰 참여자 분들의 도움으로
              <br /> 사이트가 발전합니다.
              <br /> 간단한 리뷰일지라도 등록해주시면
              <br /> 다양한 사용자에게 도움이 됩니다.
            </p>
          </div>
          <div className={styles.reviewList}>
            {recentReviewData.content.rows.map((review, index) => {
              return (
                <div className={styles.reviewBlock} key={review.id}>
                  <div className={styles.headBlock}>
                    <p className={styles.user}>{review.nickname}</p>
                    <p className={styles.time}>
                      {setDateLatest(review.createdAt)}
                    </p>
                  </div>
                  <h4
                    className={styles.storeName}
                    onClick={() =>
                      moveTargetStore(
                        review.id,
                        review.storeName,
                        review.storeLatitude,
                        review.storeLongitude,
                        review.storeAddress
                      )
                    }
                    data-testid={`moveTargetStore${index}`}
                  >
                    {review.storeName}
                  </h4>
                  <p className={styles.review}>{review.content}</p>
                  <div className={styles.hashtagList}>
                    {review.hashtag.map((hashtag) => {
                      return (
                        <div className={styles.hashtag} key={hashtag}>
                          #{hashtag}
                        </div>
                      );
                    })}
                  </div>
                </div>
              );
            })}
          </div>
        </div>
      </section>
    );
  } else {
    return <div></div>;
  }
};

export default RecentReview;
