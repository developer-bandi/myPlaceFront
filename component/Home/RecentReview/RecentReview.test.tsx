import {
  fireEvent,
  render,
  renderHook,
  screen,
  waitFor,
} from "@testing-library/react";
import configureMockStore from "redux-mock-store";
import {Provider} from "react-redux";
import axios from "axios";
import useRecentReview from "./RecentReviewHook";
import RecentReview from "./RecentReview";

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;
const recentReviewMock = {
  count: 15,
  rows: new Array(10).fill(0).map((data, index) => {
    return {
      id: index + 1,
      content: "testContent",
      createdAt: "testCreatedAt",
      storeName: "testStoreName",
      storeAddress: "testStoreAddress",
      storeLatitude: "testStoreLatitude",
      storeLongitude: "testStoreLongitude",
      nickname: "testNickname",
      hashtag: ["testHashtag"],
    };
  }),
};

describe("RecentReview Hook 테스트", () => {
  const mockStore = configureMockStore()({});
  const wrapper = ({children}: any) => (
    <Provider store={mockStore}>{children}</Provider>
  );
  describe("데이터 로드 useEffect test", () => {
    it("정상적으로 받아오는 경우", () => {
      mockedAxios.get.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: recentReviewMock,
        })
      );

      const {result} = renderHook(() => useRecentReview(), {
        wrapper,
      });

      waitFor(() => {
        expect(result.current.recentReviewData).toStrictEqual({
          content: recentReviewMock,
          loading: false,
          error: false,
        });
      });
    });

    it("에러가 발생한 경우", () => {
      mockedAxios.get.mockImplementation(() =>
        Promise.reject({
          status: 500,
          data: "error",
        })
      );
      const {result} = renderHook(() => useRecentReview(), {
        wrapper,
      });

      waitFor(() => {
        expect(result.current.recentReviewData).toStrictEqual({
          loading: false,
          error: true,
        });
      });
    });
  });
});

describe("RecentReview Presentational 테스트", () => {
  it("로딩중", () => {
    const utils = render(
      <RecentReview
        recentReviewData={{loading: true, error: false}}
        moveTargetStore={jest.fn()}
      />
    );
    screen.getByTestId("loading");
    expect(utils.container).toMatchSnapshot();
  });
  it("에러발생", () => {
    const utils = render(
      <RecentReview
        recentReviewData={{loading: false, error: true}}
        moveTargetStore={jest.fn()}
      />
    );
    expect(utils.container).toMatchSnapshot();
    screen.getByTestId("error");
  });
  it("정상 출력", () => {
    const moveTargetStoreMock = jest.fn();
    const utils = render(
      <RecentReview
        recentReviewData={{
          content: recentReviewMock,
          loading: false,
          error: false,
        }}
        moveTargetStore={moveTargetStoreMock}
      />
    );
    fireEvent.click(screen.getByTestId("moveTargetStore0"));
    screen.getByTestId("result");
    expect(utils.container).toMatchSnapshot();
    expect(moveTargetStoreMock).toBeCalledTimes(1);
  });
});
