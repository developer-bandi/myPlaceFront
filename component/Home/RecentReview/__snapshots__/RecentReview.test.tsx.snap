// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentReview Presentational 테스트 로딩중 1`] = `
<div>
  <section
    class="mainBlock"
    data-testid="loading"
  >
    <div
      class="subBlock"
    >
      <div
        class="leftBlock"
      >
        <div
          class="title"
        >
          <div
            class="loadingTitle"
          />
          <div
            class="loadingTitle"
          />
        </div>
        <div
          class="summary"
        >
          <div
            class="loadingSummary"
          />
          <div
            class="loadingSummary"
          />
          <div
            class="loadingSummary"
          />
          <div
            class="loadingSummary"
          />
        </div>
      </div>
      <div
        class="reviewList"
      >
        <div
          class="reviewBlock"
        >
          <div
            class="loadingUser"
          />
          <div
            class="loadingStoreName"
          />
          <div
            class="loadingReview"
          />
          <div
            class="loadingReview"
          />
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="loadingUser"
          />
          <div
            class="loadingStoreName"
          />
          <div
            class="loadingReview"
          />
          <div
            class="loadingReview"
          />
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="loadingUser"
          />
          <div
            class="loadingStoreName"
          />
          <div
            class="loadingReview"
          />
          <div
            class="loadingReview"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;

exports[`RecentReview Presentational 테스트 에러발생 1`] = `
<div>
  <section
    class="mainBlock"
    data-testid="error"
  >
    <p
      class="error"
    >
      에러가 발생하였습니다
    </p>
  </section>
</div>
`;

exports[`RecentReview Presentational 테스트 정상 출력 1`] = `
<div>
  <section
    class="mainBlock"
    data-testid="result"
  >
    <div
      class="subBlock"
    >
      <div
        class="leftBlock"
      >
        <h3
          class="title"
        >
          현재
          <span
            class="color"
          >
            15
          </span>
          개의
          <br />
           리뷰가 등록되어 있습니다.
        </h3>
        <p
          class="summary"
        >
          리뷰 참여자 분들의 도움으로
          <br />
           사이트가 발전합니다.
          <br />
           간단한 리뷰일지라도 등록해주시면
          <br />
           다양한 사용자에게 도움이 됩니다.
        </p>
      </div>
      <div
        class="reviewList"
      >
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore0"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore1"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore2"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore3"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore4"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore5"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore6"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore7"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore8"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
        <div
          class="reviewBlock"
        >
          <div
            class="headBlock"
          >
            <p
              class="user"
            >
              testNickname
            </p>
            <p
              class="time"
            >
              testCreatedAt
            </p>
          </div>
          <h4
            class="storeName"
            data-testid="moveTargetStore9"
          >
            testStoreName
          </h4>
          <p
            class="review"
          >
            testContent
          </p>
          <div
            class="hashtagList"
          >
            <div
              class="hashtag"
            >
              #
              testHashtag
            </div>
          </div>
        </div>
      </div>
    </div>
  </section>
</div>
`;
